README.md
# hyp-gf-coupons
Hypericum GF Coupons


change_log.txt
# 2.8.3 | 2019-01-29
	- Added the "gform_coupons_can_apply_coupon" PHP filter enabling custom logic to be used to determine if the coupon code can be applied.

# 2.8.2 | 2018-12-19
	- Fixed issue where conditional logic would generate error if the input being evaluated did not have an ID attribute (e.g. List field inputs).

# 2.8.1 | 2018-12-14
	- Added the "gform_coupons_post_delete_coupon" JS action hook which is triggered after a coupon is deleted from the field.
		gform.addAction( 'gform_coupons_post_delete_coupon', function ( code, formId ) {
		    // do stuff
		} );

# 2.8 | 2018-11-30
   - Added support for GF 2.4 conditional logic.

# 2.7 | 2018-11-02
   - Added $entry as the fourth parameter of the "gform_coupons_discount_amount" filter.
   - Added GPL to plugin header.
   - Updated the Start Date and End Date datepickers on the edit coupon page to use the YYYY-MM-DD format.
   - Updated Plugin URI and Author URI to use https.
   - Fixed Euro currency formatting of the flat amount when the edit coupon page is displayed.


# 2.6 | 2017-04-26
   - Added security enhancement.
   - Added translations.
   - Fixed code style.


# 2.5 | 2017-02-22
   - Added security enhancement.
   - Added 'isCoupon' flag when adding coupons to order summary; allows 3rd parties to more reliable identify coupons in the order summary.


# 2.4 | 2017-01-05
   - Added gform_coupons_is_valid_code to override coupon code alphanumeric checking in Gravity Forms 2.2.2.
   - Updated coupons.js to use the gform_product_total filter with a later priority.
   - Updated how default label for new fields is set.
   - Fixed wrong domain for translating the "Coupon" button.
   - Fixed an issue with the formatting of the percentage coupon amount when a decimal comma based currency is in use.


# 2.3 | 2016-07-29
   - Added Dutch translation (nl_NL).  Credit: Maarten Emmerink
   - Updated gform_coupon_script dependencies to include gform_gravityforms.
   - Updated Coupon Code validation to only allow alphanumeric characters.
   - Updated bulk actions and feed action links.
   - Fixed issue where "undefined" was output before coupon markup when applying a coupon
   - Fixed front-end JS error when attempting to remove an applied coupon if the coupon code includes special characters.


# 2.2 | 2015-07-27
   - Added GF_Field_Coupon::get_value_export() to return entry value formatted as {couponName} ({couponCode}: {discount}). Multiple codes will be comma separated e.g. $50 OFF (50OFF: -$50.00), 25% OFF (25OFF: -$10.00). Used for entry exports and some framework based add-ons since GF 1.9.12.4.
   - Added required and custom validation message settings.
   - Updated Spanish (es_ES) translation.
   - Updated minimum Gravity Forms version to 1.9.5.
   - Updated scripts and stylesheets to use minified files by default unless SCRIPT_DEBUG is active or query param gform_debug is set. e.g. domain.com/?gform_debug
   - Updated to alert if a Total field is not present when adding the Coupon field.
   - Fixed an issue with one of the field validation messages not being translatable.
   - Fixed an issue with the coupons not being included in the product info when notifications are sent after submission.
   - Fixed an issue with duplicate coupon codes being allowed to be created.
   - Fixed an issue with the coupon field where removed coupons were not removed from the hidden coupon detail input resulting in the appearance that the coupon was reapplied when returning to the page containing the coupon field.
   - Fixed an issue with the upgrade routine which caused it run even if a pre-framework version hadn't been installed.


# 2.1 | 2015-04-20
   - Added Spanish (es_ES) translation.
   - Updated gform_coupon_script deps to include gform_json.
   - Fixed a low severity security vulnerability in the admin area which could be exploited by authenticated users with form administration permissions.


# 2.0 | 2015-03-30
   - Added Addon Framework integration.
   - Added support for using the coupon field with conditional logic.
   - Added support for placeholders.
   - Added POT file.
   - Added a bulk action to the feed page for resetting the coupon usage count.
   - Added the visibility setting to the fields advanced tab.
   - Added the capability to view the plugin page.
   - Added text domain/path to header.
   - Added 'gform_coupons_discount_amount' PHP hook to support modifying the discount amount.
        add_filter( 'gform_coupons_discount_amount', 'add_logged_in_user_bonus_discount', 10, 3 );
        function add_logged_in_user_bonus_discount( $discount, $coupon, $price ) {
            if( is_user_logged_in() ) {
                $discount += 5;
            }
            return $discount;
        }
   - Added 'gform_coupons_discount_amount' JS hook.
        gform.addFilter( 'gform_coupons_discount_amount', function( discount, couponType, couponAmount, price, totalDiscount ) {
            // you would need to write your own JS-version of is_user_logged_in()
            if( is_user_logged_in() ) {
                discount += 5;
            }
            return discount;
        } );

   - Fixed an issue with the ajax request for applying coupons not functioning for logged out users.
   - Fixed an issue with feed processing which in some situations could prevent the coupon usage count being updated.
   - Fixed an issue with the field label for attribute.
   - Fixed an issue with the usage count being lost when updating an existing coupon.
   - Fixed an issue where pasting the coupon code using right-click caused the Apply button to not enable.
   - Fixed issue where existing coupon code was not saved when editing an entry.
   - Fixed issue with some translation text domain strings being misspelled.


# 1.1 | 2014-09-24
   - Updated plugin updated method so that it takes advantage of proxy manager to get around the blacklisted IP issue.

coupons.php
<?php
/*
Plugin Name: HYP Gravity Forms Coupons Add-On
Plugin URI: https://github.com/hypericumimpex/hyp-gf-coupons/
Description: Enables Gravity Forms administrators to create coupon codes that can be applied to products, services or subscriptions when used in conjunction with a payment add-on such as PayPal and Authorize.net
Version: 2.8.3
Author: rocketgenius
Author URI: https://github.com/hypericumimpex/
License: GPL-2.0+
Text Domain: gravityformscoupons
Domain Path: /languages

------------------------------------------------------------------------
Copyright 2009-2018 rocketgenius

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

define( 'GF_COUPONS_VERSION', '2.8.3' );

add_action( 'gform_loaded', array( 'GF_Coupons_Bootstrap', 'load' ), 5 );

class GF_Coupons_Bootstrap {

	public static function load() {

		if ( ! method_exists( 'GFForms', 'include_feed_addon_framework' ) ) {
			return;
		}

		require_once( 'class-gf-coupons.php' );

		GFAddOn::register( 'GFCoupons' );
	}
}

function gf_coupons() {
	return GFCoupons::get_instance();
}


js/coupons.js
function ApplyCouponCode(formId) {
    var couponCode = jQuery('#gf_coupon_code_' + formId).val();
    if (couponCode === 'undefined' || couponCode == '') {
        return;
    }

    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_spinner').show();
    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_button').prop('disabled', true);

    jQuery.post(gform_coupon_script_strings['ajaxurl'], {
            action: 'gf_apply_coupon_code',
            couponCode: couponCode,
            existing_coupons: jQuery('#gf_coupon_codes_' + formId).val(),
            formId: formId,
            total: jQuery('#gf_total_no_discount_' + formId).val()
        },

        function (response) {
            var couponInfo = jQuery.parseJSON(response);

            jQuery('#gf_coupons_container_' + formId + ' .gf_coupon_invalid').remove();
            jQuery('#gf_coupon_code_' + formId).val('');

            if (!couponInfo['is_valid']) {
                jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_info').prepend("<div class='gf_coupon_invalid'><span>" + couponInfo['invalid_reason'] + '</span></div>');
            } else {

                window['gf_coupons' + formId] = couponInfo['coupons'];

                // setting hidden field with list of coupons
                var coupon,
                    coupon_codes = '',
                    i = 0;

                for (coupon in window['gf_coupons' + formId]) {
                    if (i > 0) {
                        coupon_codes += ',';
                    }
                    coupon_codes += window['gf_coupons' + formId][coupon]['code'];

                    i++;
                }

                jQuery('#gf_coupon_codes_' + formId).val(coupon_codes).change();
                jQuery('#gf_coupons_' + formId).val(jQuery.toJSON(window['gf_coupons' + formId]));

                gformCalculateTotalPrice(formId);

            }

            jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_spinner').hide();

        }
    );
}

function GetDiscount(couponType, couponAmount, price, totalDiscount) {
    var discount;

    price = price - totalDiscount;
    if (couponType == 'flat') {
        discount = Number(couponAmount);
    } else {
        discount = price * Number((couponAmount / 100));
    }

    return gform.applyFilters('gform_coupons_discount_amount', discount, couponType, couponAmount, price, totalDiscount);
}

function PopulateDiscountInfo(price, formId) {
    var code,
        coupon,
        couponDiscount,
        couponDetails = '',
        safeCode,
        totalDiscount = 0,
        currency = new Currency(gf_global['gf_currency_config']);

    if (window['gf_coupons' + formId] === undefined) {
        window['gf_coupons' + formId] = jQuery.evalJSON(jQuery('#gf_coupons_' + formId).val());
    }

    for (code in window['gf_coupons' + formId]) {
        coupon = window['gf_coupons' + formId][code];
        couponDiscount = GetDiscount(coupon['type'], coupon['amount'], price, totalDiscount);
        totalDiscount += couponDiscount;
        safeCode = coupon.code.replace(/[^A-Za-z0-9]/g, '');

        couponDetails += '<tr class="gf_coupon_item" id="gf_coupon_' + safeCode + '"><td class="gf_coupon_name_container">' +
        '   <a href="javascript:void(0);" onclick="DeleteCoupon(\'' + coupon['code'] + '\' , \'' + formId + '\');">(x)</a>' +
        '   <span class="gf_coupon_name">' + coupon['name'] + '</span>' +
        '</td><td class="gf_coupon_discount_container">' +
        '   <span class="gf_coupon_discount">-' + currency.toMoney(couponDiscount,true) + '</span>' +
        '</td></tr>';
    }

    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_info').html('<table>' + couponDetails + '</table>');

    return totalDiscount;
}

function DisableApplyButton(formId) {
    var is_disabled = window['new_total_' + formId] == 0 || jQuery('#gf_coupon_code_' + formId).val() == '';

    if (is_disabled) {
        jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_button').prop('disabled', true);
    } else {
        jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_button').prop('disabled', false);
    }
}

gform.addFilter('gform_product_total', function (total, formId) {
    // Ignore forms that don't have a coupon field.
    if (jQuery('#gf_coupon_code_' + formId).length == 0) {
        return total;
    }

    jQuery('#gf_total_no_discount_' + formId).val(total);

    var coupon_code = gformIsHidden(jQuery('#gf_coupon_code_' + formId)) ? '' : jQuery('#gf_coupon_codes_' + formId).val(),
        has_coupon = coupon_code != '' || jQuery('#gf_coupons_' + formId).val() != '',
        new_total = total;

    if (has_coupon) {
        var total_discount = PopulateDiscountInfo(total, formId);
        new_total = total - total_discount;
    }

    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_spinner').hide();
    window['new_total_' + formId] = new_total;
    DisableApplyButton(formId);

    return new_total;
}, 50);

function DeleteCoupon(code, formId) {
    // check if coupon code is in the process of being applied
    if (jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_spinner').is(':visible')) {
        return;
    }

    var safeCode = code.replace(/[^A-Za-z0-9]/g, '');

    // removing coupon from UI
    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_' + safeCode).remove();
    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_spinner').show();
    jQuery('#gf_coupons_container_' + formId + ' #gf_coupon_button').prop('disabled', true);

    // removing coupon from coupon codes hidden input
    var codes_input = jQuery('#gf_coupon_codes_' + formId),
        coupon_codes = codes_input.val().split(','),
        index = jQuery.inArray(code, coupon_codes);

    if (index == -1) {
        return;
    }

    coupon_codes.splice(index, 1);
    codes_input.val(coupon_codes.join(',')).change();

    // removing coupon from coupon details hidden input
    if (code in window['gf_coupons' + formId]) {
        delete window['gf_coupons' + formId][code];
        jQuery('#gf_coupons_' + formId).val(jQuery.toJSON(window['gf_coupons' + formId]));
    }

    gformCalculateTotalPrice(formId);

    /**
     * Enables custom actions to be performed after the coupon is deleted.
     *
     * @since 2.8.1
     *
     * @param string code   The coupon code which was deleted.
     * @param int    formId The ID of the current form.
     */
    gform.doAction( 'gform_coupons_post_delete_coupon', code, formId );
}

// add support for conditional logic rules based on the applied coupons
gform.addFilter('gform_is_value_match', function (isMatch, formId, rule) {
    var couponField = jQuery('#field_' + formId + '_' + rule.fieldId).find('#gf_coupon_codes_' + formId);
    if (couponField.length) {
        return gf_matches_operation(couponField.val(), rule.value, rule.operator);
    }
    return isMatch;
});

// reset coupon field when hidden by conditional logic
gform.addAction('gform_post_conditional_logic_field_action', function (formId, action, targetId, defaultValues, isInit) {
    if (!isInit && action == 'hide') {
        var target = jQuery(targetId),
            coupon_items = target.find('tr.gf_coupon_item');

        if (coupon_items.length) {
            coupon_items.remove();
            target.find('input[type=hidden]').val('').change();
        }
    }
});

// Add support for GF 2.4 conditional logic by intercepting field meta when the coupon codes input is changed and
// populating correct field ID and form ID.
gform.addFilter( 'gform_field_meta_raw_input_change', function( fieldMeta, $input, event ) {

	if( $input.attr( 'id' ) && $input.attr( 'id' ).indexOf( 'gf_coupon_codes_' ) === 0 ) {
		fieldMeta.fieldId = $input.attr( 'name' ).split( '_' )[1];
		fieldMeta.formId  = $input.attr( 'id' ).split( '_' )[3];
	}

	return fieldMeta;
} );